src/System.Memory/src/System/SpanHelpers.cs

```diff
@@ -52,7 +52,36 @@ public static IntPtr Add<T>(this IntPtr start, int index)
         /// <summary>
         /// Determine if a type is eligible for storage in unmanaged memory. TODO: To be replaced by a ContainsReference() api.
         /// </summary>
-        public static bool IsReferenceFree<T>() => PerTypeLatches<T>.IsReferenceFree;
+        public static bool IsReferenceFree<T>()
+        {
+            // Under the JIT, these become constant-folded.
```

Comment:
> We already have the Type in hand

Ah, right, at that point the method is no longer generic, so JIT likely won't collapse those checks...

I'm just wondering if there is a way to improve the perf of the one-time check, while not breaking the inlining of accessing the `PerTypeLatches<T>.IsReferenceFree` field.

Something like...

```c#
internal static class SpanHelpers
{
    public static bool IsReferenceFree<T>() => PerTypeValues<T>.IsReferenceFree;

    private static bool IsReferenceFreeCore<T>()
    {
        // Under the JIT, these become constant-folded.
        if (typeof(T) == typeof(byte))
            return true;
        if (typeof(T) == typeof(sbyte))
            return true;
        if (typeof(T) == typeof(bool))
            return true;
        if (typeof(T) == typeof(char))
            return true;
        if (typeof(T) == typeof(short))
            return true;
        if (typeof(T) == typeof(ushort))
            return true;
        if (typeof(T) == typeof(int))
            return true;
        if (typeof(T) == typeof(uint))
            return true;
        if (typeof(T) == typeof(long))
            return true;
        if (typeof(T) == typeof(ulong))
            return true;
        if (typeof(T) == typeof(IntPtr))
            return true;
        if (typeof(T) == typeof(UIntPtr))
            return true;

        return IsReferenceFreeCoreSlow(typeof(T));
    }

    private static bool IsReferenceFreeCoreSlow(Type type)
    {
        if (type.GetTypeInfo().IsPrimitive) // This is hopefully the common case. All types that return true for this are value types w/out embedded references.
            return true;

        if (!type.GetTypeInfo().IsValueType)
            return false;

        // If type is a Nullable<> of something, unwrap it first.
        Type underlyingNullable = Nullable.GetUnderlyingType(type);
        if (underlyingNullable != null)
            type = underlyingNullable;

        if (type.GetTypeInfo().IsEnum)
            return true;

        foreach (FieldInfo field in type.GetTypeInfo().DeclaredFields)
        {
            if (field.IsStatic)
                continue;
            if (!IsReferenceFreeCoreSlow(field.FieldType))
                return false;
        }
        return true;
    }

    public static class PerTypeValues<T>
    {
        public static readonly bool IsReferenceFree = IsReferenceFreeCore<T>();
    }
}
```